core/src/main/scala/kafka/network/SocketServer.scala

```diff
@@ -230,12 +230,16 @@ private[kafka] abstract class AbstractServerThread(connectionQuotas: ConnectionQ
 
   /**
    * Close the connection identified by `connectionId` and decrement the connection count.
+   * The channel will be immediately removed from the selector's `channels` or `closingChannels`
+   * and no further disconnect notifications will be sent for this channel by the selector.
    */
   def close(selector: KSelector, connectionId: String): Unit = {
     val channel = selector.channel(connectionId)
-    if (channel != null) {
+    val openChannel = selector.channel(connectionId)
+    val openOrClosingChannel = if (openChannel != null) openChannel else selector.closingChannel(connectionId)
```

Comment:
Maybe we can use `openOrClosingChannel` here and any other place that follows a similar pattern.

Reply:
Done.